Lib/ast.py

```diff
@@ -368,6 +368,67 @@ def walk(node):
         yield node
 
 
+def compare(
+    a,
+    b,
+    /,
+    *,
+    compare_types=True,
+    compare_fields=True,
+    compare_attributes=False,
+):
+    """
+    Compares recursively given two ast nodes. If *compare_types* is False, the
+    field values won't be checked whether they belong to same type or not. If
+    *compare_fields* is True, members of `_fields` attribute on both node's type
+    will be checked. If *compare_attributes* is True, members of `_attributes`
+    attribute on both node's will be compared.
+    """
+
+    def _compare(a, b):
+        if compare_types and type(a) is not type(b):
+            return False
+        elif isinstance(a, AST):
+            return compare(
+                a,
+                b,
+                compare_attributes=compare_attributes,
+                compare_fields=compare_fields,
+                compare_types=compare_types,
+            )
+        elif isinstance(a, list):
+            if len(a) != len(b):
+                return False
```

Comment:
I think you're assuming here that b is also a list, but if compare_types is False is might not be?

Reply:
In theory, if you didn't alter the AST by your hand, a sequence field is always a list. So when comparing 2 different ASTs, if the field 'X' of AST 'A' has the type list, then the field 'X' of AST 'B' should be also a list (this is only for sequence types thought, not for optionals).

Reply:
compare_types is about the types of Python objects, while the list here is about the Python list used for fields that have repeated values.